### PR TITLE
NIFI-7467 Refactored S2S peer selection logic

### PR DESCRIPTION
**Pull request by alopresto**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

* Removed list structure for peer selection as it was unnecessary and often wasteful (most clusters are 3 - 7 nodes, the list was always 128 elements).
* Changed integer percentages to double to allow for better normalization.
* Removed 80% cap on remote peers as it was due to legacy requirements.
* Added unit tests for non-deterministic distribution calculations.
* Added unit tests for edge cases due to rounding errors, single valid remotes, unbalanced clusters, and peer queue consecutive selection tracking.
* Migrated all legacy PeerSelector unit tests to new API.
* Removed unused System time manipulation as tests no longer need it.
* Added class-level Javadoc to PeerSelector.
* Removed S2S details request replication, as the responses were not being merged, which led to incorrect ports being returned and breaking S2S peer retrieval.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by alopresto** [comment]
@thenatog I believe this PR resolves the error you and I both got during testing of PR 4263 where if a secure cluster had multiple nodes running on the same physical/virtual server using different ports, Site-to-site would fail to retrieve the peers because the certificate SAN did not match. What was happening was the request for S2S details was being replicated to all nodes, and whichever answered first used its ports, but often did not match the specific node which was queried.

**Comment by thenatog** [comment]
Very cool, will review soon

**Comment by alopresto** [comment]
I ran the GitHub Actions checks twice. The problem is with downloading a dependency for another module from a remote repository. Running the build on my local machine completes successfully. 

```
[ERROR] Exception calling reference $projectsSortedByOrganization at META-INF/DEPENDENCIES.vm[line 34, column 33]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-remote-resources-plugin:1.7.0:process (process-resource-bundles) on project nifi-hive_1_1-processors: Execution process-resource-bundles of goal org.apache.maven.plugins:maven-remote-resources-plugin:1.7.0:process failed: Error resolving project artifact: Failure to transfer eigenbase:eigenbase-properties:pom:1.1.4 from https://repository.apache.org/content/repositories/releases was cached in the local repository, resolution will not be reattempted until the update interval of apache-repo has elapsed or updates are forced. Original error: Could not transfer artifact eigenbase:eigenbase-properties:pom:1.1.4 from/to apache-repo (https://repository.apache.org/content/repositories/releases): Transfer failed for https://repository.apache.org/content/repositories/releases/eigenbase/eigenbase-properties/1.1.4/eigenbase-properties-1.1.4.pom 503 Service Unavailable for project eigenbase:eigenbase-properties:jar:1.1.4 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :nifi-hive_1_1-processors
##[error]Process completed with exit code 1.
```

File: nifi-commons/nifi-site-to-site-client/pom.xml [hunk 1]
```diff
@@ -76,6 +76,15 @@
             <artifactId>httpasyncclient</artifactId>
             <version>4.1.4</version>
         </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-commons/nifi-site-to-site-client/pom.xml:83]: I think this was added during exploration around the test logging failing but is not needed now. Removed.

File: nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java [hunk 2]
```diff
@@ -16,353 +16,559 @@
...
-        for (int i = 0; i < peerList.size(); i++) {
-            final long idx = peerIndex.getAndIncrement();
-            final int listIndex = (int) (idx % peerList.size());
-            peerStatus = peerList.get(listIndex);
+        final double totalWeights = sumMapValues(orderedPeerStatuses);
+        logger.debug("Determining next available peer ({} peers with total weight {})", orderedPeerStatuses.keySet().size(), totalWeights);
+
+        final double random = Math.random() * 100;
+        logger.debug("Generated random value {}", random);
+        if (random > totalWeights) {
```
[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java:425]: This is a much cleaner solution. I was trying during the iterative changes to stay close to the prior design and did not think of this. Thanks.

File: nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java [hunk 3]
```diff
@@ -16,353 +16,559 @@
...
+                if (isPenalized(peerStatus)) {
+                    logger.debug("{} is penalized; will not communicate with this peer", peerStatus);
+                } else {
+                    return peerStatus;
+                }
             }
         }
 
-        logger.debug("{} All peers appear to be penalized; returning null", this);
+        logger.debug("Did not select a peer; r {}, t {}, w {}", random, threshold, orderedPeerStatuses.values());
```
[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java:444]: Same as the above comment, I was trying not to modify the existing logic where possible, but I think the idea to remove the penalized peers before selecting one is the best path forward. Implementing this.

File: nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java [hunk 4]
```diff
@@ -16,353 +16,559 @@
...
-            ioe.addSuppressed(lastFailure);
+            // Persist the fetched peer statuses
+            PeerStatusCache peerStatusCache = new PeerStatusCache(statuses, System.currentTimeMillis(), peerStatusProvider.getTransportProtocol());
+            persistPeerStatuses(peerStatusCache);
+            logger.info("Successfully refreshed peer status cache; remote group consists of {} peers", statuses.size());
+        } catch (Exception e) {
+            warn(logger, eventReporter, "Unable to refresh remote group peers due to: {}", e.getMessage());
+            if (logger.isDebugEnabled() && e.getCause() != null) {
+                Throwable cause = e.getCause();
+                while ((cause = cause.getCause()) != null) {
```
[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java:568]: The reason this was introduced was because the exceptions thrown here are often very long (hundreds of lines of stacktrace) and occur very often (if the TLS config is incorrect, this can be thrown many times per second, making the log difficult/impossible to read). I've simplified this to just print the stacktrace if debug is enabled, but I think there is still room for improvement down the line.

File: nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerStatusProvider.java [hunk 5]
```diff
@@ -52,12 +51,25 @@
...
+     * Fetch peer statuses from a remote NiFi cluster.
+     * Implementation of this method should fetch peer statuses from the node
+     * represented by the passed PeerDescription using its transport protocol.
+     *
+     * @param peerDescription a bootstrap node or one of query-able nodes lastly fetched successfully
+     * @param communicator the communication mechanism provider can be provided directly
+     * @return Remote peer statuses
+     * @throws IOException thrown when it fails to fetch peer statuses of the remote cluster from the specified peer
+     */
+    Set<PeerStatus> fetchRemotePeerStatuses(final PeerDescription peerDescription, SiteToSiteCommunicator communicator) throws IOException;
```
[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerStatusProvider.java:71]: I am not opposed to genericizing the class, but I think it introduces more complexity as every instance would need to determine in advance what communicator implementation it accepted. 

I would argue that while compile-time checks are obviously better than runtime, passing a new anonymous instantiation of a marker interface is "allowed" but "an intentionally bad decision" like many other bad decisions which can occur in code. 

I'm going to look at other options to allow reusability of the specific client instance for performance, perhaps without exposing that functionality directly to the caller.

[the remark this answers is not in the document]
Comment (alopresto) [reply nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerStatusProvider.java:71]: This was only used once internal to `HttpClient`, so I just removed it from the interface and cleaned up the unsupported implementation.

**Comment by alopresto** [comment]
Thanks Mark. I replied to the comments and integrated most of what you had suggested (the exception being the complete removal of the marker interface).
